src/BloomBrowserUI/bookEdit/js/bloomEditing.js

```diff
@@ -653,30 +653,56 @@ $(document).ready(function() {
     if (typeof CKEDITOR === "undefined") return;  // this happens during unit testing
     CKEDITOR.disableAutoInline = true;
 
+    // Map from ckeditor id strings to the top offset of the corresponding edit box.
+    var mapCkeditBoxTop = new Object();
+
     // attach ckeditor to the contenteditable="true" class="bloom-content1"
     $('div.bloom-page').find('.bloom-content1[contenteditable="true"]').each(function() {
 
         var ckedit = CKEDITOR.inline(this);
 
+        // Record the top of the edit box for use later.  A bug in ckeditor
+        // misplaces the format bar vertically when the edit box starts out
+        // empty and the format bar starts out hidden.  This bookkeeping helps
+        // to work around that bug.
+        var boxTop0 = this.offsetTop;
+        // We need to go three levels to get the offset used by the format bar.
+        if (this.offsetParent) {
+            boxTop0 += this.offsetParent.offsetTop;
+            if (this.offsetParent.offsetParent)
+                boxTop0 += this.offsetParent.offsetParent.offsetTop;
+        }
+        mapCkeditBoxTop[ckedit.id] = boxTop0;
+
         // show or hide the toolbar when the text selection changes
         ckedit.on('selectionCheck', function(evt) {
             var editor = evt['editor'];
-            var rng = editor.getSelection().getRanges()[0];
-            var show = (rng.startOffset !== rng.endOffset);
+            // Length of selected text is more reliable than compariing
+            // endpoints of the first range.  Mozilla can return multiple
+            // ranges with the first one being empty.
+            var selection = editor.getSelection();
+            var textSelected = selection.getSelectedText();
+            var show = (textSelected.length > 0);
             var bar = $('body').find('.' + editor.id);
-            bar.css('top', bar.data('top'));
             show ? bar.show() : bar.hide();
+
+            // Move the format bar on the screen if needed.
+            // (Note that offsets are not defined if it's not visible.)
+            if (show) {
+                var barTop = bar.offset().top;
+                var boxTop = mapCkeditBoxTop[editor.id];
+                if (boxTop - barTop < 5) {
+                    var barLeft = bar.offset().left;
+                    var barHeight = bar.height();
+                    bar.offset({ top: boxTop - barHeight, left: barLeft });
+                }
+            }
```

Comment:
I think the map is updated whenever the page is shown.  I've tried it on a variety of pages and page layouts.  I'm not sure how to bring about "text editing changes the position of the text box" however, or how to get the scroll position to change.  Our text boxes don't seem to scroll or change position no matter what I type, or how much I type.